File: Systems-for-Design-and-Implementation/Labs/Lab12/ProiectJava/TransportNetworking/src/main/java/transport/network/rpcprotocol/TransportClientRpcReflectionWorker.java
package transport.network.rpcprotocol;

import transport.model.Reservation;
import transport.model.ReservationManager;
import transport.model.Trip;
import transport.services.ITransportObserver;
import transport.services.ITransportServices;
import transport.services.TransportException;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.net.Socket;
import java.sql.SQLOutput;

public class TransportClientRpcReflectionWorker implements Runnable, ITransportObserver {
    private ITransportServices server;
    private Socket connection;

    private ObjectInputStream input;
    private ObjectOutputStream output;
    private volatile boolean connected;
    public TransportClientRpcReflectionWorker(ITransportServices server, Socket connection) {
        this.server = server;
        this.connection = connection;
        try{
            output=new ObjectOutputStream(connection.getOutputStream());
            output.flush();
            input=new ObjectInputStream(connection.getInputStream());
            connected=true;
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void run() {
        while(connected){
            try {
                Object request=input.readObject();
                Response response=handleRequest((Request)request);
                if (response!=null){
                    sendResponse(response);
                }
            } catch (IOException e) {
                e.printStackTrace();
            } catch (ClassNotFoundException e) {
                e.printStackTrace();
            }
            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
        try {
            input.close();
            output.close();
            connection.close();
        } catch (IOException e) {
            System.out.println("Error "+e);
        }
    }

    public void reservationAdded(Reservation reservation) {
        Response resp = new Response.Builder().type(ResponseType.NEW_RESERVATION).data(reservation).build();
        System.out.println("Reservation added: " + reservation);
        try {
            sendResponse(resp);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private static Response okResponse=new Response.Builder().type(ResponseType.OK).build();

    private Response handleRequest(Request request){
        Response response=null;
        String handlerName="handle"+(request).type();
        System.out.println("HandlerName "+handlerName);
        try {
            Method method=this.getClass().getDeclaredMethod(handlerName, Request.class);
            response=(Response)method.invoke(this,request);
            System.out.println("Method "+handlerName+ " invoked");
        } catch (NoSuchMethodException e) {
            e.printStackTrace();
        } catch (InvocationTargetException e) {
            e.printStackTrace();
        } catch (IllegalAccessException e) {
            e.printStackTrace();
        }

        return response;
    }

    private Response handleLOGIN(Request request){
        System.out.println("Login Request ... " + request.type());
        ReservationManager reservationManager = (ReservationManager)request.data();
        try {
            server.login(reservationManager, this);
            return okResponse;
        } catch (TransportException e) {
            connected=false;
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }

    private Response handleSEARCH_RESERVATION_MANAGER_BY_NAME(Request request){
        System.out.println("Search Reservation Manager By Name Request ...");
        String name=(String)request.data();
        try {
            return new Response.Builder().type(ResponseType.SEARCH_RESERVATION_MANAGER_BY_NAME).data(server.searchReservationManagerByName(name)).build();
        } catch (TransportException e) {
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }

    private Response handleGET_ALL_TRIPS(Request request){
        System.out.println("Get All Trips Request ...");
        try {
            return new Response.Builder().type(ResponseType.GET_ALL_TRIPS).data(server.getAllTrips()).build();
        } catch (TransportException e) {
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }

    private Response handleSEARCH_TRIP_BY_ID(Request request) {
        System.out.println("Search Trip By Id Request ...");
        Long id = (Long) request.data();
        try {
            return new Response.Builder().type(ResponseType.SEARCH_TRIP_BY_ID).data(server.searchTripById(id)).build();
        } catch (TransportException e) {
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }

    private Response handleADD_RESERVATION(Request request){
        System.out.println("Add Reservation Request ...");
        Reservation reservation = (Reservation)request.data();
        new Thread(() -> {
            try {
                server.addReservation(reservation);
            } catch (TransportException e) {
                e.printStackTrace();
            }
        }).start();
        return okResponse;
    }

    private Response handleGET_RESERVATIONS_BY_TRIP(Request request){
        System.out.println("Get Reservations By Trip Request ...");
        Trip trip = (Trip)request.data();
        try {
            return new Response.Builder().type(ResponseType.GET_RESERVATIONS_BY_TRIP).data(server.getReservationsByTrip(trip)).build();
        } catch (TransportException e) {
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }

    private Response handleGET_SEATS_BY_RESERVATION(Request request){
        System.out.println("Get Seats By Reservation Request ...");
        Long reservationId = (Long)request.data();
        try {
            return new Response.Builder().type(ResponseType.GET_SEATS_BY_RESERVATION).data(server.getSeatsByReservation(reservationId)).build();
        } catch (TransportException e) {
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }

    private Response handleGET_SEATS_BY_TRIP(Request request) {
        System.out.println("Get Seats By Trip Request ...");
        Trip trip = (Trip)request.data();
        try {
            return new Response.Builder().type(ResponseType.GET_SEATS_BY_TRIP).data(server.getSeatsByTrip(trip)).build();
        } catch (TransportException e) {
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }

    private Response handleLOGOUT(Request request){
        System.out.println("Logout Request...");
        ReservationManager reservationManager=(ReservationManager)request.data();
        try {
            server.logout(reservationManager, this);
            connected=false;
            return okResponse;
        } catch (TransportException e) {
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }

    private void sendResponse(Response response) throws IOException{
        System.out.println("sending response "+response);
        synchronized (output) {
            output.writeObject(response);
            output.flush();
        }
    }
}

File: Systems-for-Design-and-Implementation/Labs/Lab8/ProiectCSharp/TransportNetworking/transport.network/rpcprotocol/TransportClientRpcReflectionWorker.cs
using System;
using System.IO;
using System.Net.Sockets;
using System.Reflection;
using System.Runtime.Serialization.Formatters.Binary;
using System.Threading;
using Newtonsoft.Json;
using TransportModel.transport.model;
using TransportServices.transport.services;
using JsonSerializer = System.Text.Json.JsonSerializer;

namespace TransportNetworking.transport.network.rpcprotocol
{
    public class TransportClientRpcReflectionWorker : ITransportObserver
    {
        private ITransportServices server;
        private Socket connection;

        private StreamReader input;
        private StreamWriter output;
        private volatile bool connected;

        public TransportClientRpcReflectionWorker(ITransportServices server, Socket connection)
        {
            this.server = server;
            this.connection = connection;
            try
            {
                NetworkStream networkStream = new NetworkStream(connection);
                output = new StreamWriter(networkStream);
                input = new StreamReader(networkStream);
                connected = true;
            }
            catch (IOException e)
            {
                Console.WriteLine(e.StackTrace);
            }
        }

        public void Run()
        {
            while (connected)
            {
                try
                {
                    string line = input.ReadLine();
                    Request request = JsonConvert.DeserializeObject<Request>(line!);
                    //BinaryFormatter formatter = new BinaryFormatter();
                    //object request = formatter.Deserialize(input.BaseStream);
                    Response response = HandleRequest(request);
                    if (response != null)
                    {
                        SendResponse(response);
                    }
                }
                catch (IOException e)
                {
                    Console.WriteLine(e.StackTrace);
                }

                Thread.Sleep(1000);
            }

            try
            {
                input.Close();
                output.Close();
                connection.Close();
            }
            catch (IOException e)
            {
                Console.WriteLine("Error " + e);
            }
        }

        public void ReservationAdded(Reservation reservation)
        {
            Response resp = new Response.Builder().Type(ResponseType.NEW_RESERVATION).Data(reservation).Build();
            Console.WriteLine("Reservation added: " + reservation);
            try
            {
                SendResponse(resp);
            }
            catch (IOException e)
            {
                Console.WriteLine(e.StackTrace);
            }
        }

        private static Response okResponse = new Response.Builder().Type(ResponseType.OK).Build();

        private Response HandleRequest(Request request)
        {
            Response response = null;
            string handlerName = "Handle" + request.Type;
            Console.WriteLine("HandlerName " + handlerName);
            try
            {
                MethodInfo method = this.GetType().GetMethod(handlerName, new Type[] { typeof(Request) });
                if (method != null)
                {
                    response = (Response)method.Invoke(this, new object[] { request });
                    Console.WriteLine("Method " + handlerName + " invoked");
                }
                else
                {
                    Console.WriteLine("Method " + handlerName + " not found");
                }
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }

            return response;
        }

        public Response HandleLOGIN(Request request)
        {
            Console.WriteLine("Login Request ... " + request.Type + " " + request.Data);
            ReservationManager reservationManager = request.Data as ReservationManager;
            if (reservationManager == null)
            {
                Console.WriteLine("Error: ReservationManager is null");
                return new Response.Builder().Type(ResponseType.ERROR).Data("ReservationManager is null").Build();
            }
            try
            {
                Console.WriteLine("Reservation Manager in HandleLOGIN: " + reservationManager);
                ReservationManager reservationManager2 = server.Login(reservationManager, this);
                Console.WriteLine("Login successful! (Worker)");
                return new Response(){Data = reservationManager2};
            }
            catch (Exception e)
            {
                Console.WriteLine("Exception during login: " + e.Message);
                Console.WriteLine(e.StackTrace);
                connected = false;
                return new Response.Builder().Type(ResponseType.ERROR).Data(e.Message).Build();
            }
        }

        public Response HandleSEARCH_RESERVATION_MANAGER_BY_NAME(Request request)
        {
            Console.WriteLine("Search Reservation Manager By Name Request ...");
            string name = (string)request.Data;
            try
            {
                return new Response.Builder().Type(ResponseType.SEARCH_RESERVATION_MANAGER_BY_NAME)
                    .Data(server.SearchReservationManagerByName(name)).Build();
            }
            catch (TransportException e)
            {
                return new Response.Builder().Type(ResponseType.ERROR).Data(e.Message).Build();
            }
        }

        public Response HandleGET_ALL_TRIPS(Request request)
        {
            Console.WriteLine("Get All Trips Request ...");
            try
            {
                return new Response.Builder().Type(ResponseType.GET_ALL_TRIPS).Data(server.GetAllTrips()).Build();
            }
            catch (TransportException e)
            {
                return new Response.Builder().Type(ResponseType.ERROR).Data(e.Message).Build();
            }
        }

        public Response HandleSEARCH_TRIP_BY_ID(Request request)
        {
            Console.WriteLine("Search Trip By Id Request ...");
            long id = (long)request.Data;
            try
            {
                return new Response.Builder().Type(ResponseType.SEARCH_TRIP_BY_ID).Data(server.SearchTripById(id)).Build();
            }
            catch (TransportException e)
            {
                return new Response.Builder().Type(ResponseType.ERROR).Data(e.Message).Build();
            }
        }

        public Response HandleADD_RESERVATION(Request request)
        {
            Console.WriteLine("Add Reservation Request ...");
            Reservation reservation = (Reservation)request.Data;
            new Thread(() =>
            {
                try
                {
                    server.AddReservation(reservation);
                }
                catch (TransportException e)
                {
                    Console.WriteLine(e.StackTrace);
                }
            }).Start();
            return okResponse;
        }

        public Response HandleGET_RESERVATIONS_BY_TRIP(Request request)
        {
            Console.WriteLine("Get Reservations By Trip Request ...");
            Trip trip = (Trip)request.Data;
            try
            {
                return new Response.Builder().Type(ResponseType.GET_RESERVATIONS_BY_TRIP).Data(server.GetReservationsByTrip(trip)).Build();
            }
            catch (TransportException e)
            {
                return new Response.Builder().Type(ResponseType.ERROR).Data(e.Message).Build();
            }
        }

        /*public Response HandleGET_SEATS_BY_RESERVATION(Request request)
        {
            Console.WriteLine("Get Seats By Reservation Request ...");
            long reservationId = (long)request.Data();
            try
            {
                return new Response.Builder().Type(ResponseType.GET_SEATS_BY_RESERVATION).Data(server.GetSeatsByReservation(reservationId)).Build();
            }
            catch (TransportException e)
            {
                return new Response.Builder().Type(ResponseType.ERROR).Data(e.Message).Build();
            }
        }

        public Response HandleGET_SEATS_BY_TRIP(Request request)
        {
            Console.WriteLine("Get Seats By Trip Request ...");
            long tripId = (long)request.Data();
            try
            {
                return new Response.Builder().Type(ResponseType.GET_SEATS_BY_TRIP).Data(server.GetSeatsByTripId(tripId)).Build();
            }
            catch (TransportException e)
            {
                return new Response.Builder().Type(ResponseType.ERROR).Data(e.Message).Build();
            }
        }*/

        public Response HandleLOGOUT(Request request)
        {
            Console.WriteLine("Logout Request...");
            ReservationManager reservationManager = (ReservationManager)request.Data;
            try
            {
                server.Logout(reservationManager, this);
                connected = false;
                return okResponse;
            }
            catch (TransportException e)
            {
                return new Response.Builder().Type(ResponseType.ERROR).Data(e.Message).Build();
            }
        }

        private void SendResponse(Response response)
        {
            Console.WriteLine("sending response " + response);
            lock (output)
            {
                string responseString = JsonConvert.SerializeObject(response);
                output.WriteLine(responseString);
                output.Flush();
                //BinaryFormatter formatter = new BinaryFormatter();
                /*using (MemoryStream ms = new MemoryStream())
                {
                    formatter.Serialize(ms, response);
                    byte[] buffer = ms.ToArray();
                    output.Write(buffer, 0, buffer.Length);
                    output.Flush();
                }*/
            }
        }
    }
}